_source/_posts/2022-08-31-state-of-java-project-loom.md

```diff
@@ -0,0 +1,261 @@
+
+## What is Project Loom?
+
+> Project Loom aims to drastically reduce the effort of writing, maintaining, and observing high-throughput concurrent applications that make the best use of available hardware.
+>
+> — Ron Pressler (Tech lead, Project Loom)
+
+OS threads are at the core of Java's concurrency model and have a very mature ecosystem around them, but they also come with some drawbacks and are expensive computationally. Let's look at the two most common use cases for concurrency and the drawbacks of the current Java concurrency model in these cases.
+
+One of the most common concurrency use cases is serving requests over the wire using a server. For this, the most preferred approach is the thread-per-request model, where a separate thread handles each request. Throughput of such systems can be explained using [Little's law](https://en.wikipedia.org/wiki/Little%27s_law), which states that in a **stable system**, the average concurrency (number of requests concurrently processed by the server), **L** is equal to the throughput (average rate of requests), **λ**, times the latency (average duration of processing each request), **W**. With this, you can derive that throughput equals average concurrency divided by latency (**λ = L/W**).
+
+So in a thread-per-request model, the throughput will be limited by the number of OS threads available, which depends on the number of physical cores/threads available on the hardware. To work around this, you have to use shared thread pools or asynchronous concurrency, both of which have their drawbacks. Thread pools have many limitations, like thread leaking, deadlocks, resource thrashing, etc. Asynchronous concurrency means you must adapt to a more complex programming style and handle data races carefully. There are also chances of memory leaks, thread locking, etc.
+
+Another common use case is parallel processing or multi-threading. You must ensure thread synchronization when executing a parallel task distributed over multiple threads to avoid data races. The implementation becomes even more fragile and puts a lot more responsibility on the developer to ensure there are no issues like thread leaks and cancellation delays.
+
+Project Loom aims to fix these issues in the current concurrency model by introducing two new features: virtual threads and structured concurrency.
+
+### Virtual Threads
+
+> Java 19 is scheduled to be released in September 2022, and Virtual threads will be a preview feature. Yayyy!
+
+[Virtual threads](https://openjdk.org/jeps/425) are lightweight threads that are not tied to OS threads but are managed by the JVM. They are suitable for thread-per-request programming styles without having the limitations of OS threads. You can create millions of virtual threads without affecting throughput. This is quite similar to coroutines, like [Goroutines](https://go.dev/tour/concurrency/1), made famous by the Go programming language.
```

Comment:
```suggestion
[Virtual threads](https://openjdk.org/jeps/425) are lightweight threads that are not tied to OS threads but are managed by the JVM. They are suitable for thread-per-request programming styles without having the limitations of OS threads. You can create millions of virtual threads without affecting throughput. This is quite similar to coroutines, like [goroutines](https://go.dev/tour/concurrency/1), made famous by the Go programming language.
```
since it isn't capitalized in the Go docs, we don't need to either...